pkg/dbconn/metadatalock.go

```diff
@@ -52,28 +51,34 @@ func NewMetadataLock(ctx context.Context, dsn string, lockName string, logger lo
 	getLock := func() error {
 		// https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html#function_get-lock
 		var answer int
-		stmt := sqlescape.MustEscapeSQL("SELECT GET_LOCK(%?, %?)", lockName, getLockTimeout.Seconds())
+		// Using the table name alone entails a maximum lock length and leads to conflicts
+		// between different tables with the same name in different schemas.
+		// We use the schema name and table name to create a unique lock name with a hash.
+		// The hash is truncated to 8 characters to avoid the maximum lock length.
+		// bizarrely_long_schema_name.thisisareallylongtablenamethisisareallylongtablename60charac ==>
+		// bizarrely_long_schem.thisisareallylongtablenamethisis-66fec116
+		stmt := sqlescape.MustEscapeSQL("SELECT GET_LOCK( concat(left(%?,20),'.',left(%?,32),'-',left(sha1(concat(%?,%?)),8)), %?)", table.SchemaName, table.TableName, table.SchemaName, table.TableName, getLockTimeout.Seconds())
 		if err := dbConn.QueryRowContext(ctx, stmt).Scan(&answer); err != nil {
 			return fmt.Errorf("could not acquire metadata lock: %s", err)
 		}
 		if answer == 0 {
 			// 0 means the lock is held by another connection
 			// TODO: we could lookup the connection that holds the lock and report details about it
-			return fmt.Errorf("could not acquire metadata lock: %s, lock is held by another connection", lockName)
+			return fmt.Errorf("could not acquire metadata lock for %s.%s, lock is held by another connection", table.SchemaName, table.TableName)
 		} else if answer != 1 {
 			// probably we never get here, but just in case
-			return fmt.Errorf("could not acquire metadata lock: %s, GET_LOCK returned: %d", lockName, answer)
+			return fmt.Errorf("could not acquire metadata lock for %s.%s, GET_LOCK returned: %d", table.SchemaName, table.TableName, answer)
 		}
 		return nil
 	}
 
 	// Acquire the lock or return an error immediately
 	// We only Infof the initial acquisition.
-	logger.Infof("attempting to acquire metadata lock: %s", lockName)
+	logger.Infof("attempting to acquire metadata lock for %s.%s", table.SchemaName, table.TableName)
 	if err = getLock(); err != nil {
 		return nil, err
 	}
-	logger.Infof("acquired metadata lock: %s", lockName)
+	logger.Infof("acquired metadata lock: %s.%s", table.SchemaName, table.TableName)
```

Comment:
May be worth logging `lockName` too in one of the logs somewhere

Reply:
It's not easy to do that, because I'm computing the lock name server-side inside the GET_LOCK function call. So for now let's leave it like this and if in the future we have a real need to log the lock name, I think we could just add another MySQL round-trip to compute the lock name and then use it in GET_LOCK.